core/common/src/Buffer.kt

```diff
@@ -80,119 +81,100 @@ public class Buffer : Source, Sink {
-        if (limit - pos < 4L) {
-            return (
-                    readByte() and 0xff shl 24
-                            or (readByte() and 0xff shl 16)
-                            or (readByte() and 0xff shl 8)
-                            or (readByte() and 0xff)
-                    )
-        }
-
-        val data = segment.data
-        val i = (
-                data[pos++] and 0xff shl 24
-                        or (data[pos++] and 0xff shl 16)
-                        or (data[pos++] and 0xff shl 8)
-                        or (data[pos++] and 0xff)
-                )
+        val segment = head
+        if (segment == null) {
+            throwEof(4)
+            return 0 // unreachable
+        }
+        val segmentSize = segment.size
+        if (segmentSize < 4) {
+            // If the short is split across multiple segments, delegate to readShort().
+            require(4)
+            if (segmentSize == 0) {
+                recycleHead()
+                return readInt()
+            }
+            return (readShort().toInt() shl 16 or (readShort().toInt() and 0xffff))
+        }
+        val v = segment.readInt()
         sizeMut -= 4L
-
-        if (pos == limit) {
+        if (segmentSize == 4) {
             recycleHead()
-        } else {
-            segment.pos = pos
         }
-
-        return i
+        return v
     }
 
     override fun readLong(): Long {
-        require(8)
-
-        val segment = head!!
-        var pos = segment.pos
-        val limit = segment.limit
-
-        // If the long is split across multiple segments, delegate to readInt().
-        if (limit - pos < 8L) {
-            return (
-                    readInt() and 0xffffffffL shl 32
-                            or (readInt() and 0xffffffffL)
-                    )
-        }
-
-        val data = segment.data
-        val v = (
-                data[pos++] and 0xffL shl 56
-                        or (data[pos++] and 0xffL shl 48)
-                        or (data[pos++] and 0xffL shl 40)
-                        or (data[pos++] and 0xffL shl 32)
-                        or (data[pos++] and 0xffL shl 24)
-                        or (data[pos++] and 0xffL shl 16)
-                        or (data[pos++] and 0xffL shl 8)
-                        or (data[pos++] and 0xffL)
-                )
-        this.sizeMut -= 8L
-
-        if (pos == limit) {
+        val segment = head
+        if (segment == null) {
+            throwEof(8)
+            return 0 // unreachable
+        }
+        val segmentSize = segment.size
+        if (segmentSize < 8) {
+            // If the short is split across multiple segments, delegate to readInt().
+            require(8)
+            if (segmentSize == 0) {
+                recycleHead()
+                return readLong()
+            }
+            return (readInt().toLong() shl 32 or (readInt().toLong() and 0xffffffffL))
```

Comment:
Is this way of reading exactly faster than using Unsafe API?

It turns out that in the worst case, we will fall down the call stack by an additional 3 frames down

Reply:
Frankly, I didn't optimize the cross-segment read/write scenarios as they should not occur as frequently as reads/writes fitting inside a single segment.

As the current implementation should not make it worse, I suggest investigating and optimizing (if needed) these sites separately.